tests/libs/common/common_tests.cpp

```diff
@@ -248,4 +245,178 @@ ring_buffer_api_test_helper(
+    std::function<void(int)> generate_event,
+    bool doubled_data)
+{
+    std::vector<std::vector<char>> records = expected_records;
+
+    // Add a couple of interleaved messages to the records vector.
+    std::string message = "Interleaved message #1";
+    std::vector<char> record(message.begin(), message.end());
+    record.push_back('\0');
+    records.push_back(record);
+    record[record.size() - 2] = '2';
+    records.push_back(record);
+
+    // Perf buffer event callback context.
+    std::unique_ptr<perf_buffer_test_context_t> context = std::make_unique<perf_buffer_test_context_t>();
+    context->test_event_count = PERF_BUFFER_TEST_EVENT_COUNT + 2;
+    context->doubled_data = doubled_data;
+
+    context->records = &records;
+
+    // Generate events prior to subscribing for ring buffer events.
+    for (int i = 0; i < PERF_BUFFER_TEST_EVENT_COUNT / 2; i++) {
+        generate_event(i);
+    }
+
+    // Write an interleaved message to the perf buffer map.
+    REQUIRE(ebpf_perf_event_array_map_write(perf_buffer_map, message.c_str(), message.length() + 1) == EBPF_SUCCESS);
+
+    // Get the std::future from the promise field in perf buffer event context, which should be in ready state
+    // once notifications for all events are received.
+    auto perf_buffer_event_callback = context->perf_buffer_event_promise.get_future();
+
+    // Create a new perf buffer manager and subscribe to perf buffer events.
+    // The notifications for the events that were generated before should occur after the subscribe call.
+    context->perf_buffer = perf_buffer__new(
+        perf_buffer_map,
+        16,
+        (perf_buffer_sample_fn)perf_buffer_test_event_handler,
+        (perf_buffer_lost_fn)perf_buffer_test_lost_event_handler,
+        context.get(),
+        nullptr);
+    REQUIRE(context->perf_buffer != nullptr);
+
+    // Generate more events, post-subscription.
+    for (int i = PERF_BUFFER_TEST_EVENT_COUNT / 2; i < PERF_BUFFER_TEST_EVENT_COUNT; i++) {
+        generate_event(i);
+    }
+
+    // Write another interleaved message to the perf buffer map.
+    message[message.length() - 1] = '2';
+    REQUIRE(ebpf_perf_event_array_map_write(perf_buffer_map, message.c_str(), message.length() + 1) == EBPF_SUCCESS);
+    // Wait for event handler getting notifications for all PERF_BUFFER_TEST_EVENT_COUNT events.
+    bool test_completed = perf_buffer_event_callback.wait_for(10s) == std::future_status::ready;
+    CAPTURE(context->matched_entry_count, context->lost_entry_count, context->test_event_count);
+    REQUIRE(test_completed == true);
+    REQUIRE((context->matched_entry_count + context->lost_entry_count) == context->test_event_count);
+
+    // Mark the event context as canceled, such that the event callback stops processing events.
+    context->canceled = true;
+
+    // Release the raw pointer such that the final callback frees the callback context.
+    perf_buffer_test_context_t* raw_context = context.release();
```

Comment:
This is no longer the case. Keep it in the smart pointer which should get destroyed after it falls out of scope.